src/ln/peer_handler.rs

```diff
@@ -1097,7 +1098,95 @@ impl<Descriptor: SocketDescriptor> PeerManager<Descriptor> {
 			}
 		};
 	}
-}
+
+	/// insure we recieved pong message from all peers or disconnect the peers then ping all peers
+	pub fn check_peer(&mut self)-> Result<(), PeerHandleError>{
```

Comment:
On the high-level you should iterate on peer list at once for both sending peers, like 

```
for peer in peers { 
        // if time to ping, schedule a ping on peer
       // if pong is wrong, disconnect peer
}
```

I think we should have consts against which we take actions like UNLATCH_NEXT_PING or WAIT_PONG_UNTIL_BEFORE_DISCONNECT and a new field per-peer like `ping_tracker`, every call to check_peer we increment. When receiving a valid pong in `do_read_event` you reset this field to 0.

Reply:
I am responding a little bit later than I wanted (sorry), but I do appreciate the suggestions as they are quite helpful :).
I really like adding the ping_tracker field to Peer and that is more or less what Matt suggested, I am a little bit unsure if we need the constants though?
currently I am thinking the logic may look something like this:

check_peers{
read_event() // interestingly I don't think we need to call process_events() after this one, here we are checking for the 
// Pong message in the pending outbound buffer and resetting ping_tracker to 0 if there is one.
loop through peers in PeerManager{
if ping_tracker > 0 then disconnect then disconnect_event
ping_peers() // this function will take a Peer and push a ping message to the pending outbound buffer, call //process_events() and then it will increment the ping_tracker by one.
}
}

Reply:
IMO the constants are needed, you want a value to check against to know when ping has timeout. It could be a value passed to check_peers() too.

I think you don't need to include `read_event` or `process_event` as they are more general function driving the whole P2P stack which should be decided to be called by the client. Instead check received pong message directly in  `do_read_event` (msg_type 19). Same for `process_event` write directly in `pending_outbound_buffer` and call `do_attempt_write_data`